diff_cover/report_generator.py

```diff
@@ -173,13 +173,22 @@ def _diff_violations(self):
         To make this efficient, we cache and reuse the result.
         """
         if not self._diff_violations_dict:
+            violations = None
+            src_paths_changed = self._diff.src_paths_changed()
+            try:
+                # Use violations_batch() if it's implemented.
+                violations = self._violations.violations_batch(src_paths_changed)
+            except NotImplementedError:
+                pass
             self._diff_violations_dict = {
                 src_path: DiffViolations(
-                    self._violations.violations(src_path),
+                    violations.get(src_path, [])
+                    if violations is not None
```

Comment:
so the violations reporter is supposed to cache these results so each is only processed once.

The function you are calling `violations_batch` could also be described as `warm_cache`.

Does that sound right to you? The reason I wanna rename this is because I want to hint that the implementor should be saving the results so we dont have to remember the state later (basically remove this if check)

basically....

Im wondering if you would go with renaming `violations_batch` to `warm_cache` and then removing the `if violations is not None` and just calling `violations` as is (under the assumption that the implementor has saved the results so the lookup is quick)

Reply:
My assumption in checking for ` if violations is not None` is that it will be `None` if the function is not implemented. In that case, wouldn't we need to call `violations()`?

Bigger picture: For SQLFluff, `violations_batch()` is not a warm cache. Rather, SQLFluff is a fairly slow-running tool that is often used to process hundreds or even thousands of files. If `diff-quality` calls SQLFluff one file at a time, there is no opportunity to use parallel processing. Adding a function like `violations_batch()` provides more opportunity for the implementor to use parallel processing or other clever strategies to improve performance. It can definitely be used for caching as well.

Does this change your thoughts on the above?

It may be helpful (in terms of code clarity) to **remove** the line `violations = None` and change the subsequent code to:
```
            try:
                violations = self._violations.violations_batch(src_paths_changed)
            except NotImplementedError:
                violations = {src_path: self._violations.violations(src_path)}
            self._diff_violations_dict = {
                src_path: DiffViolations(
                    violations.get(src_path, []),
                    ...
```

This ensures that `violations` is always defined and not `None`.

Reply:
My proposed code above doesn't work "as is" because it uses `src_path` outside the loop. I'll try to do something similar. I was trying to avoid looping over the list of files twice, but maybe that's the best option.